### PR TITLE
Let the user settings hook just use the first defined value

### PR DESCRIPTION
**Pull request by rafpaf**
This pull request adds a new option to to the `useUserSetting` hook: just use the first defined value. This helps avoid unnecessary re-renders, which avoids race conditions.

In the `useUserSetting` function, an additional parameter `shouldUseFirstDefinedValue` was added in this branch. If this parameter is true, only the first value retrieved from the API is used. This parameter is optional and set to false by default.

### CODE REVIEW

**Comment by rafpaf** [comment]
* **#42574** <a href="https://app.graphite.dev/github/pr/metabase/metabase/42574?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#42605** <a href="https://app.graphite.dev/github/pr/metabase/metabase/42605?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rafpaf and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by iethree — CHANGES_REQUESTED** [review]
1) This really needs tests. This is essentially a caching problem which is really hard, and tests will help make sure it works as we expect, and clearly document those expectations.
2) I'm not sure memoization like this is the right approach:

Here, we say, for the purposes of this hook, ignore all API values other than the first one.  I think this is confusing to reason about, especially when multiple components are subscribing to the same setting.

I think what we really want here is aggressively optimistic updates - where we immediately update the value in the redux store and then _don't update the redux store_ when the API request completes. That way we get completely synchronous behavior client side for these user settings, and hopefully save them server side, but it's no big deal if they don't. This also lets us avoid the weird back/forth race condition possibilities.  I also think this should be the default behavior, maybe with the option to turn it off. WDYT?

File: frontend/src/metabase/common/hooks/use-setting/use-setting.ts [hunk 1]
```diff
@@ -38,5 +46,8 @@ export const useUserSetting = <T extends keyof UserSettings>(
     () => _.debounce(setter, debounceTimeout, debounceOnLeadingEdge),
     [setter, debounceTimeout, debounceOnLeadingEdge],
   );
-  return [currentValue, shouldDebounce ? debouncedSetter : setter];
+  return [
+    shouldMemoize ? memoizedValue : currentValue,
+    shouldDebounce ? debouncedSetter : setter,
```
Comment (sloansparger) [inline frontend/src/metabase/common/hooks/use-setting/use-setting.ts:51]: If this were going to go through, I think there should be a setter that does something. You should be able to at the very least have your value update to the setter's value.

**Review by sloansparger — CHANGES_REQUESTED** [review]
I'd like to understand a bit more why this was needed. If useSetting is slow, we should fix it for everyone instead of opting in. If you have a singular use case for this, I'd opt for a more composable solution like:

```
const CheapToRerender = () => {
  const [setting] = useSetting('my-setting')
  const staleSetting = useFirstDefinedValue(setting);

  return (
     <ExpensiveToRerender thing={setting.attr} />
   )
}
```

File: frontend/src/metabase/common/hooks/use-setting/use-setting.ts (hunk 1 above)
Comment (rafpaf) [reply frontend/src/metabase/common/hooks/use-setting/use-setting.ts:51]: Not sure I follow

Comment (sloansparger) [reply frontend/src/metabase/common/hooks/use-setting/use-setting.ts:51]: const [value, set] = useSetting('thing', { shouldMemoize: true });
console.log(value) // 'a'

// somewhere else
set('b');

// next render
const [value, set] = useSetting('thing', { shouldMemoize: true });
console.log(value) // 'a' <-- still a even though we set it to b

**Comment by rafpaf** [comment]
> I'd like to understand a bit more why this was needed. 

The prop makes it possible to remove a race condition in the Browse models filter: <a href="https://github.com/metabase/metabase/pull/42574" target="_blank">Avoid race condition in Browse models filter</a>. The race condition could arise for any similar UI that tracks a value in the API, so a general solution seemed useful

**Comment by rafpaf** [comment]
> 1. This really needs tests. This is essentially a caching problem which is really hard, and tests will help make sure it works as we expect, and clearly document those expectations.

Agreed. Will write some tests.

> 2. I'm not sure memoization like this is the right approach:
> 
> Here, we say, for the purposes of this hook, ignore all API values other than the first one. I think this is confusing to reason about, especially when multiple components are subscribing to the same setting.
> 
> I think what we really want here is aggressively optimistic updates - where we immediately update the value in the redux store and then _don't update the redux store_ when the API request completes. That way we get completely synchronous behavior client side for these user settings, and hopefully save them server side, but it's no big deal if they don't. This also lets us avoid the weird back/forth race condition possibilities. I also think this should be the default behavior, maybe with the option to turn it off. WDYT?

I like this idea!

**Comment by rafpaf** [comment]
Closing for now. Will return to this if the priority is determined to be higher.

**Pull request closed without merging**